internal/sandbox/mapped_dir.go

```diff
@@ -77,142 +80,196 @@ type openMappedDir struct {
-		// affected. Thus, switch to scaffold directory if present.
-		if scaffoldOK {
-			return scaffoldChild, nil
+	if underlyingPath, isMapped := d.UnderlyingPath(); isMapped {
+		childPath := filepath.Join(underlyingPath, name)
+		fileInfo, err := os.Lstat(childPath)
+		if err != nil {
+			return nil, fuseErrno(err)
 		}
-		return nil, fuseErrno(statErr)
-	}
 
-	if child := d.baseChildFromFileInfo(fileInfo); child != nil {
-		d.baseChildren[name] = child
+		child := getOrCreateNode(childPath, fileInfo, d.BaseNode.writable)
+		d.children[name] = child
 		return child, nil
 	}
-	return scaffoldChild, nil
+
+	// We don't know about the entry and we don't have a backing directory, so there is nothing
+	// else we can return.
+	return nil, fuse.ENOENT
 }
 
 // Lookup looks for a particular directory/file in all the children of a given
 // directory.
 func (d *MappedDir) Lookup(_ context.Context, name string) (fs.Node, error) {
-	return d.lookup(name)
+	d.mu.Lock()
+	defer d.mu.Unlock()
+
+	return d.lookupUnlocked(name)
 }
 
 // Dirent returns the directory entry corresponding to the directory.
 func (d *MappedDir) Dirent(name string) fuse.Dirent {
 	return fuse.Dirent{
-		Inode: d.Inode(),
+		Inode: d.Inode(), // No need to lock: inode numbers are immutable.
 		Name:  name,
 		Type:  fuse.DT_Dir,
 	}
 }
 
-// ReadDirAll lists all files/directories inside a directory.
-func (o *openMappedDir) ReadDirAll(context.Context) ([]fuse.Dirent, error) {
+// readEntries obtains the list of raw directory entries from the opened directory.
+func (o *openMappedDir) readEntries() ([]os.FileInfo, error) {
 	o.mu.Lock()
+	defer o.mu.Unlock()
+
+	if o.nativeDir == nil {
+		return []os.FileInfo{}, nil
+	}
+
 	if o.needRewind {
 		o.nativeDir.Seek(0, os.SEEK_SET)
 	}
+	var err error
 	dirents, err := o.nativeDir.Readdir(-1)
 	if err != nil {
-		o.mu.Unlock()
 		return nil, fuseErrno(err)
 	}
 	o.needRewind = true
-	o.mu.Unlock()
+	return dirents, nil
 
-	done := make(map[string]bool)
+}
+
+// ReadDirAll lists all files/directories inside a directory.
+func (o *openMappedDir) ReadDirAll(context.Context) ([]fuse.Dirent, error) {
+	dirents, err := o.readEntries()
+	if err != nil {
+		return nil, err
+	}
 
 	o.dir.mu.Lock()
 	defer o.dir.mu.Unlock()
-	dirEntries := make([]fuse.Dirent, 0, len(o.dir.mappedChildren)+len(dirents)+len(o.dir.scaffoldDirs))
-	for name, node := range o.dir.mappedChildren {
-		dirEntries = append(dirEntries, node.Dirent(name))
-		done[name] = true
-	}
-	for _, fileInfo := range dirents {
-		if done[fileInfo.Name()] {
-			continue
-		}
-		if child := o.dir.baseChildFromFileInfo(fileInfo); child != nil {
-			o.dir.baseChildren[fileInfo.Name()] = child
-			dirEntries = append(dirEntries, child.Dirent(fileInfo.Name()))
-			done[fileInfo.Name()] = true
-		}
-	}
-	for name, node := range o.dir.scaffoldDirs {
-		if !done[name] {
+
+	dirEntries := make([]fuse.Dirent, 0, len(o.dir.children)+len(dirents))
```

Comment:
dirents vs dirEntries is a bit confusing here

How about renaming readEntries() to readNativeEntries()

And renaming dirents to nativeDirEntries

And renaming dirEntries to fuseDirEntries

Or something like that... ? I'm not sure, probably you can think of better names :)

Reply:
Done.